### ISSUE TITLE
Transpiled template literals behave incorrectly

### ISSUE THREAD

**Issue by lowr** [microsoft/TypeScript#39744]
**TypeScript Version:**  3.9.7, nightly

**Search Terms:**

template literals, template strings, ES5, ES3

**Code**

```ts
class C {
    toString() {
        return "toString()";
    }
    valueOf() {
        return "valueOf()";
    }
}

const c = new C;
const s = `${c}`;
console.log(s);
```

**Expected behavior:**

According to ES2015 (and later) spec, the result of expressions in template literals are converted to primitives first by calling `toString()` and then by `valueOf()` (unless there is `[Symbol.toPrimitive]()`, but I'll just ignore it here as `Symbol`s wouldn't get transpiled). `s` therefore evaluates to `"toString()"`.

**Actual behavior:**

`s` evaluates to `"valueOf()"`.

The above code is compiled to the following when the target is ES5 or ES3. As `+` operator attempts to convert its operands to primitives first by calling `valueOf()` and then by `toString()`, which is the opposite of the expected behavior.

```js
"use strict";
var C = /** @class */ (function () {
    function C() {
    }
    C.prototype.toString = function () {
        return "toString()";
    };
    C.prototype.valueOf = function () {
        return "valueOf()";
    };
    return C;
}());
var c = new C;
var s = "" + c;
console.log(s);
```

FWIW, Babel compiles the above code to [this](https://babeljs.io/en/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAEDC0G8BQ1XQC4HsDK6BOAlgHYDmAFAJSIpq14Cm6ArnkdAERa6GmXsDcNVAF8h0AG5gQTegHkAZpWq06jFm3aTpcxRQFjRopMExEI6aMGgBeaEXoB3OIJNmLMWwAMAJAmDDPF1MITBB6ADoQTHIICn4gA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ces2015%2Ctypescript%2Cenv&prettier=false&targets=&version=7.10.5&externalPlugins=) unless loose transformation is enabled. It makes use of `String.prototype.concat()` and behaves as expected.

**Playground Link:** [Link](https://www.typescriptlang.org/play/?target=1&ts=4.0.0-dev.20200725&ssl=1&ssc=1&pln=12&pc=16#code/MYGwhgzhAEDC0G8BQ1XQC4HsDK6BOAlgHYDmAFAJSIpq14Cm6ArnkdAERa6GmXsDcNVAF8h0AG5gQTegHkAZpWq06jFm3aTpcxRQFjRopMExEI6aMGgBeaEXoB3OIJNmLMWwAMAJAmDDPF1MITBB6ADoQTHIICn4gA)

**Related Issues:** 

#30239: Although OP's suggestion is not related, some people commenting on the issue seem to confront this issue (like [this comment](https://github.com/microsoft/TypeScript/issues/30239#issuecomment-548721302))

**Comment by DanielRosenwasser** [comment]
Is this something you ran into in a real-world app?

**Comment by lowr** [comment]
Yes, I had some trouble when I was trying to interact with Discord API using [discord.js](https://github.com/discordjs/discord.js).

The library defines a class `User`, which extends another class `Base`. `User` overrides `toString()` so its instances can be serialized into a format Discord API specifies using template literals (see [example](https://discord.js.org/#/docs/main/stable/class/User?scrollTo=toString) on the library's documentation) while `Base` overrides `valueOf()` presumably for its internal uses. I tried to serialize a `User` instance with template literals as the example shows, which didn't work as expected since the transpiled code would call `valueOf()` from the `Base` class instead of `toString()` from the `User` class.

**Comment by gilmoreorless** [comment]
As a heads-up, this behaviour will cause problems with the upcoming [`Temporal`](https://github.com/tc39/proposal-temporal/) proposal. Currently some of the `Temporal` objects are [specified to throw an error in their `valueOf()` methods](https://tc39.es/proposal-temporal/#sec-temporal.plaindate.prototype.valueof).

There's a discussion about TypeScript's transpilation behaviour for template literals at https://github.com/tc39/proposal-temporal/issues/1681

**Comment by ljharb** [comment]
This is a spec compliance issue that babel also had to deal with; it’s worth fixing.

It also comes up when interpolating Symbols.

**Comment by justingrant** [comment]
> As a heads-up, this behaviour will cause problems with the upcoming [`Temporal`](https://github.com/tc39/proposal-temporal/) proposal. Currently some of the `Temporal` objects are [specified to throw an error in their `valueOf()` methods](https://tc39.es/proposal-temporal/#sec-temporal.plaindate.prototype.valueof).
> 
> There's a discussion about TypeScript's transpilation behaviour for template literals at [tc39/proposal-temporal#1681](https://github.com/tc39/proposal-temporal/issues/1681)

Yep. As a champion of the (currently Stage 3) Temporal proposal, this issue is concerning because it will cause divergence in behavior between spec-compliant interpolation of Temporal instances (which will succeed) and TSC-polyfilled interpolation (which will throw). 

Given that TS transpilation is sometimes done via babel and sometimes done via TSC, I can see this creating a lot of confusion among developers, esp. those who are less familiar with the bundler/transpilation plumbing that underlies most modern apps.

Also, a common use case for interpolation of Temporal instances will be console.log statements, because logs often have timestamps or durations displayed. It'd be a bummer to have logging code that works great in JS suddenly start crashing when porting that code to TypeScript.

**Comment by ljharb** [comment]
Totally agree. Hopefully typescript, which has a tc39 presence, can match the spec before it becomes a problem.

**Comment by justingrant** [comment]
@andrewbranch would be great to see this in 4.5 final release if possible. The first (flagged) implementations of Temporal are starting to be released in JS engines, e.g. Safari just shipped the first tech preview with a partial Temporal implementation. And production polyfills are maturing. Would be great to have this fixed in TS before there's too much in-the-wild usage.  Thanks!
